src/environments.jl

```diff
@@ -4,79 +4,238 @@ struct Environments
+end
+
+function getGuessInds(A::fPEPS, prev_mps, col::Int)
+    # guess should have the indices of A going left/right that prev_mps does not have
+    Ny, Nx = size(A)
+    A_prev_common  = [commonind(A[row, col], prev_mps[row]) for row in 1:Ny]
+    hori_A_inds    = [inds(A[row, col], "Link, r") for row in 1:Ny]
+    hori_prev_inds = [inds(prev_mps[row], "Link, r") for row in 1:Ny]
+    double_hori_A  = [IndexSet(hori_A_inds[row]..., prime(hori_A_inds[row])...) for row in 1:Ny]
+    A_prev_unique  = [setdiff(double_hori_A[row], hori_prev_inds[row]) for row in 1:Ny]
+    return A_prev_unique
+end
+
+function initGuessRandomMPS(A::fPEPS, prev_mps, col::Int, chi::Int)
+    @timeit "randomMPS" begin
+        Ny, Nx = size(A)
+        A_prev_unique  = getGuessInds(A, prev_mps, col)
+        hori_cmbs      = Vector{ITensor}(undef, Ny)
+        hori_cis       = Vector{Index}(undef, Ny)
+        @inbounds for row in 1:Ny
+            cmb            = combiner(A_prev_unique[row]..., tags="r$row,CMB,Site")
+            hori_cmbs[row] = cmb 
+            hori_cis[row]  = combinedind(cmb) 
+        end
+        guess = randomMPS(hori_cis, chi)
+        @inbounds for row in 1:Ny
+            guess[row] *= hori_cmbs[row]
+        end
+    end
+    return guess
+end
+
+function initGuessRandomITensor(A::fPEPS, prev_mps, col::Int, chi::Int)
+    @timeit "randomITensor" begin
+        Ny, Nx = size(A)
+        A_prev_unique  = getGuessInds(A, prev_mps, col)
+        guess = MPS(Ny)
+        up_inds        = [Index(chi, "Link,u,c$col,r$row") for row in 1:Ny-1]
+        @inbounds for row in 1:Ny
+            row_inds = nothing
+            if row == 1
+                row_inds = IndexSet(A_prev_unique[row]..., up_inds[row])
+            elseif 1 < row < Ny
+                row_inds = IndexSet(A_prev_unique[row]..., up_inds[row-1], up_inds[row])
+            else
+                row_inds = IndexSet(A_prev_unique[row]..., up_inds[row-1])
+            end
+            #guess[row] = is_cu ? randomCuITensor(row_inds) : randomITensor(row_inds) 
+            guess[row] = randomITensor(row_inds) 
+        end
+    end
+    return guess
+end
+
+function constructEnvsAbove(A::fPEPS, guess::MPS, prev_mps::Vector{<:ITensor}, ops::Vector{ITensor}, col::Int, sweep::Int)
+    is_cu  = is_gpu(A)
+    Ny, Nx = size(A)
+    order  = isodd(sweep) ? (1:Ny) : reverse(1:Ny)
+    dummy  = is_cu ? cuITensor(1.0) : ITensor(1.0)
+    Envs_above = Vector{ITensor}(undef, Ny+1)
+    Envs_above[end]  = is_cu ? cuITensor(1.0) : ITensor(1.0)
+    @inbounds for (ii, env_row) in enumerate(reverse(order))
+        tmp = A[env_row, col]
+        tmp *= ops[env_row]
+        tmp *= prev_mps[env_row]
+        next_row = isodd(sweep) ? env_row + 1 : env_row - 1
+        Env_above  = next_row > 0 ? copy(Envs_above[next_row]) : dummy 
+        Env_above *= tmp
+        Env_above *= dag(prime(A[env_row, col]))
+        Env_above *= guess[env_row]
```

Comment:
I believe switching the above two lines makes this function D^5 instead of D^6

Reply:
Tried this and it actually made scaling worse! Reverted back to the old order.

Reply:
Ok, good that you checked. We'll have to come back to some of this after doing timings for empirical scaling. But right now everything in this part of the code looks to be no more than D^6 and that's the most important thing!